hibernate-core/src/test/java/org/hibernate/orm/test/mapping/converted/converter/ConvertedAttributesTypecheckTest.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+package org.hibernate.orm.test.mapping.converted.converter;
+
+import java.time.Duration;
+import java.util.Set;
+
+import org.hibernate.testing.orm.junit.DomainModel;
+import org.hibernate.testing.orm.junit.Jira;
+import org.hibernate.testing.orm.junit.SessionFactory;
+import org.hibernate.testing.orm.junit.SessionFactoryScope;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import jakarta.persistence.AttributeConverter;
+import jakarta.persistence.Convert;
+import jakarta.persistence.Converter;
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.Id;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Marco Belladelli
+ */
+@DomainModel( annotatedClasses = ConvertedAttributesTypecheckTest.TestEntity.class )
+@SessionFactory
+@Jira( "https://hibernate.atlassian.net/browse/HHH-17693" )
+public class ConvertedAttributesTypecheckTest {
+	@BeforeAll
+	public void setUp(SessionFactoryScope scope) {
+		scope.inTransaction( session -> session.persist( new TestEntity(
+				Set.of( "one", "two" ),
+				"123",
+				String.valueOf( Duration.ofDays( 3 ).toMillis() )
+		) ) );
+	}
+
+	@AfterAll
+	public void tearDown(SessionFactoryScope scope) {
+		scope.inTransaction( session -> session.createMutationQuery( "delete from TestEntity" ).executeUpdate() );
+	}
+
+	@Test
+	public void testLikeOnConvertedString(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			final TestEntity result = session.createQuery(
+					"from TestEntity where convertedString like 'one%'",
+					TestEntity.class
+			).getSingleResult();
+			assertThat( result.getConvertedString() ).contains( "one" );
+		} );
+	}
+
+	@Test
+	public void testUnaryExpressionOnConvertedNumber(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			final String result = session.createQuery(
+					"select -convertedNumber from TestEntity",
+					String.class
+			).getSingleResult();
+			assertThat( result ).isEqualTo( "-123" );
+		} );
+	}
+
+	@Test
+	public void testFromDurationExpressionOnConvertedDuration(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			final String result = session.createQuery(
+					"select convertedDuration by day from TestEntity",
+					String.class
+			).getSingleResult();
+			assertThat( Long.parseLong( result ) ).isEqualTo( Duration.ofDays( 3 ).toMillis() );
+		} );
+	}
```

Comment:
> I guess that the idea is that the converted field has an underlying JDBC type that's different to its Java type, and so we can do things with it that we could not do to the Java type

Yes that was the idea, I implemented this tests specifically to ensure my changes to `TypecheckUtil#assertNumeric` (only used in `SemanticQueryBuilder#visitUnaryExpression`) and `TypecheckUtil#assertDuration` (only used in `SemanticQueryBuilder#visitFromDurationExpression`) worked as expected with converted attributes. 

> But that doesn't justify saying that the converter is applied recursively to all containing expressions

In both of these tests, the selected expression (`SqmUnaryOperation` or `SqmByUnit`) produces results typed using the underlying basic path, which in this case is converted thus creating a result that is also converted the same way the property itself would be. I also ran some tests and this was the case since Hibernate 6.1.

If you think this is not correct it would be definitely possible to change this behavior to account for converters and adapt these tests to new expectations, but it would be a shift in logic.

Reply:
Yeah, so what I'm saying is I don't think it's intuitive at all that this converter acts transitively on the result of an operator expression.

Nor do I see it as something well-defined. The `by day` operator has the effect of changing the type of the thing it's applied to (it converts an expression of type `Duration` to something of type `Integer`. So then what does it mean to apply a converter from `Duration` to `String` to the `Integer` result of that operation? The same objection applies to any other operator which can change the type of its operand.

So the thing here is that this was all reasonably well-defined when we were type checking comparison expressions with `==` or whatever, because you just checked the database-level values for equality, and the result type was always `Boolean`. A comparison operator never evaluates to one of its operands.

But for these operations, that's not the case, so we have to be able to answer the question of what is the result type of the operation.

Reply:
(Not at all sure what to do here.)

Reply:
This one is completely fine, by the way:

    from TestEntity where convertedString like 'one%'

since it's a comparison operator.